File: source/meta.ts
/**
 * @copyright Copyright (c) 2026 GUIHO Technologies as represented by Cristóvão GUIHO. All Rights Reserved.
 */

import { readDirectory as readdir, statPath as stat } from './runtime/fs.js'
import { basename, dirname, joinPath as join, relativePath as relative, resolvePath as resolve } from './runtime/path.js'
import type {
  XDocsConfig,
  XDocsFrontmatter,
  XDocsMetaDescriptor,
  XDocsMetaDocument,
  XDocsMetaFilters,
  XDocsMetaScanOptions,
  XDocsMetaScanResult,
  XDocsMetadata,
} from './types.js'
import { XDocsError } from './errors.js'
import { isPlainMarkdownDocument, isXDocsDescriptorFile } from './discovery.js'
import { parseFrontmatterObject, readFrontmatterFromFile, validateMetadata } from './metadata.js'

export {
  collectMetaErrors,
  scanMetadata,
}

type MarkdownDocumentReference = {
  path: string
  relativePath: string
  directory: string
  name: string
}

/** Scan xdocs descriptors and optional companion docs by reading frontmatter only. */
const scanMetadata = async (config: XDocsConfig, options: XDocsMetaScanOptions = {}): Promise<XDocsMetaScanResult> => {
  const targetPath = resolve(config.cwd, options.targetPath ?? config.cwd)
  const targetStat = await stat(targetPath).catch(() => null)

  if (!targetStat) throw new XDocsError(`Metadata target does not exist: ${relative(config.cwd, targetPath) || '.'}`)
  if (!targetStat.isDirectory()) throw new XDocsError(`Metadata target must be a directory: ${relative(config.cwd, targetPath) || '.'}`)

  const descriptors: XDocsMetaDescriptor[] = []
  const documentsByDirectory = new Map<string, MarkdownDocumentReference[]>()

  await walkMetadataDirectory(targetPath, config, async (filePath) => {
    if (isXDocsDescriptorFile(filePath)) {
      descriptors.push(await parseMetaDescriptor(filePath, config.cwd))
      return
    }

    if (isPlainMarkdownDocument(filePath)) {
      addDocumentReference(documentsByDirectory, {
        path: filePath,
        relativePath: relative(config.cwd, filePath),
        directory: dirname(filePath),
        name: basename(filePath),
      })
    }
  })

  await enrichMetaDescriptors(descriptors, documentsByDirectory, Boolean(options.includeDocuments))

  const filters = createMetaFilters(options)
  const filteredDescriptors = filterDescriptors(descriptors, filters, Boolean(options.includeDocuments))
  const errors = collectMetaErrors({ descriptors: filteredDescriptors })

  return {
    root: config.cwd,
    targetPath: relative(config.cwd, targetPath) || '.',
    includeDocuments: Boolean(options.includeDocuments),
    strict: Boolean(options.strict),
    filters,
    descriptors: filteredDescriptors,
    errors,
  }
}

/** Collect validation errors from a metadata scan or descriptor list. */
const collectMetaErrors = (input: Pick<XDocsMetaScanResult, 'descriptors'>): string[] => {
  const errors: string[] = []

  for (const descriptor of input.descriptors) {
    for (const error of descriptor.errors) errors.push(`${descriptor.relativePath}: ${error}`)
    for (const document of descriptor.documents) {
      for (const error of document.errors) errors.push(`${document.relativePath}: ${error}`)
    }
  }

  return errors
}

const parseMetaDescriptor = async (filePath: string, cwd: string): Promise<XDocsMetaDescriptor> => {
  const errors: string[] = []

  if (basename(filePath).toLowerCase() === '.xdocs.md') {
    errors.push('Invalid xdocs descriptor filename. Use a named file such as "authentication.xdocs.md"; ".xdocs.md" is only the extension.')
  }

  const frontmatterText = await readFrontmatterFromFile(filePath)
  let frontmatter: XDocsFrontmatter | null = null
  let metadata: XDocsMetadata | null = null

  if (!frontmatterText) {
    errors.push('Missing YAML frontmatter.')
  } else {
    const parsed = parseFrontmatterObject(frontmatterText)
    frontmatter = parsed.frontmatter
    errors.push(...parsed.errors)

    if (frontmatter) {
      const result = validateMetadata(frontmatter)
      if (result.valid) {
        metadata = result.metadata
      } else {
        errors.push(...result.errors)
      }
    }
  }

  return {
    path: filePath,
    relativePath: relative(cwd, filePath),
    directory: dirname(filePath),
    subject: metadata?.subject ?? stringField(frontmatter, 'subject'),
    valid: metadata !== null && errors.length === 0,
    frontmatter,
    metadata,
    documents: [],
    errors,
  }
}

const parseMetaDocument = async (
  document: MarkdownDocumentReference,
  expectedOwner: string | null,
): Promise<XDocsMetaDocument> => {
  const errors: string[] = []
  const frontmatterText = await readFrontmatterFromFile(document.path)
  let frontmatter: XDocsFrontmatter | null = null

  if (!frontmatterText) {
    errors.push('Missing YAML frontmatter.')
  } else {
    const parsed = parseFrontmatterObject(frontmatterText)
    frontmatter = parsed.frontmatter
    errors.push(...parsed.errors)
    if (frontmatter) errors.push(...validateCompanionFrontmatter(frontmatter, expectedOwner))
  }

  return {
    ...document,
    owner: stringField(frontmatter, 'owner'),
    valid: frontmatter !== null && errors.length === 0,
    frontmatter,
    errors,
  }
}

const enrichMetaDescriptors = async (
  descriptors: XDocsMetaDescriptor[],
  documentsByDirectory: Map<string, MarkdownDocumentReference[]>,
  includeDocuments: boolean,
): Promise<void> => {
  const descriptorsByDirectory = new Map<string, XDocsMetaDescriptor[]>()

  for (const descriptor of descriptors) {
    const siblings = descriptorsByDirectory.get(descriptor.directory) ?? []
    siblings.push(descriptor)
    descriptorsByDirectory.set(descriptor.directory, siblings)
  }

  for (const siblings of descriptorsByDirectory.values()) {
    if (siblings.length <= 1) continue
    for (const descriptor of siblings) {
      descriptor.errors.push('Multiple xdocs descriptors found in this directory. Keep exactly one named "*.xdocs.md" file per directory.')
    }
  }

  for (const descriptor of descriptors) {
    const documents = documentsByDirectory.get(descriptor.directory) ?? []
    validateDocumentReferences(descriptor, documents)
    if (includeDocuments) descriptor.documents = await parseAssociatedDocuments(descriptor, documents)
    descriptor.valid = descriptor.metadata !== null && descriptor.errors.length === 0
  }
}

const parseAssociatedDocuments = async (
  descriptor: XDocsMetaDescriptor,
  documents: MarkdownDocumentReference[],
): Promise<XDocsMetaDocument[]> => {
  if (!descriptor.metadata) return []

  const declaredDocuments = descriptor.metadata.documents
  const associatedDocuments = documents.filter((document) => Object.hasOwn(declaredDocuments, document.name))
  const result: XDocsMetaDocument[] = []

  for (const document of associatedDocuments) {
    result.push(await parseMetaDocument(document, descriptor.metadata.subject))
  }

  return result
}

const validateDocumentReferences = (
  descriptor: XDocsMetaDescriptor,
  documents: MarkdownDocumentReference[],
): void => {
  if (!descriptor.metadata) return

  const actualDocuments = new Set(documents.map((document) => document.name))
  const declaredDocuments = descriptor.metadata.documents

  for (const document of documents) {
    if (Object.hasOwn(declaredDocuments, document.name)) continue
    descriptor.errors.push(`Undocumented Markdown document: "${document.name}" must be listed in the "documents" metadata map.`)
  }

  for (const name of Object.keys(declaredDocuments)) {
    if (!isPlainMarkdownDocumentName(name)) {
      descriptor.errors.push(`Invalid document entry: "${name}" must be a sibling plain "*.md" filename, not an xdocs descriptor or path.`)
      continue
    }

    if (!actualDocuments.has(name)) {
      descriptor.errors.push(`Missing Markdown document: "${name}" is listed in metadata but does not exist beside the descriptor.`)
    }
  }
}

const validateCompanionFrontmatter = (frontmatter: XDocsFrontmatter, expectedOwner: string | null): string[] => {
  const errors: string[] = []

  for (const field of ['name', 'purpose', 'description', 'created', 'owner']) {
    if (typeof frontmatter[field] !== 'string' || frontmatter[field].length === 0) {
      errors.push(`Missing or invalid "${field}" field. Expected a non-empty string.`)
    }
  }

  for (const field of ['flags', 'tags', 'keywords']) {
    const value = frontmatter[field]
    if (!Array.isArray(value) || value.some((entry) => typeof entry !== 'string')) {
      errors.push(`Missing or invalid "${field}" field. Expected an array of strings.`)
    }
  }

  const created = stringField(frontmatter, 'created')
  if (created && !/^\d{4}-\d{2}-\d{2}$/.test(created)) {
    errors.push('Invalid "created" field. Expected YYYY-MM-DD.')
  }

  const owner = stringField(frontmatter, 'owner')
  if (expectedOwner && owner && owner !== expectedOwner) {
    errors.push(`Invalid "owner" field. Expected "${expectedOwner}".`)
  }

  return errors
}

const filterDescriptors = (
  descriptors: XDocsMetaDescriptor[],
  filters: XDocsMetaFilters,
  includeDocuments: boolean,
): XDocsMetaDescriptor[] => {
  if (!hasFilters(filters)) return descriptors

  return descriptors
    .map((descriptor) => {
      const documents = includeDocuments
        ? descriptor.documents.filter((document) => documentMatchesFilters(document, filters))
        : []

      return { ...descriptor, documents }
    })
    .filter((descriptor) => descriptorMatchesFilters(descriptor, filters) || descriptor.documents.length > 0)
}

const descriptorMatchesFilters = (descriptor: XDocsMetaDescriptor, filters: XDocsMetaFilters): boolean => {
  if (filters.owner && !sameValue(descriptor.subject, filters.owner)) return false
  if (filters.tag && !arrayFieldIncludes(descriptor.frontmatter, 'tags', filters.tag)) return false
  if (filters.keyword && !arrayFieldIncludes(descriptor.frontmatter, 'keywords', filters.keyword)) return false
  return true
}

const documentMatchesFilters = (document: XDocsMetaDocument, filters: XDocsMetaFilters): boolean => {
  if (filters.owner && !sameValue(document.owner, filters.owner)) return false
  if (filters.tag && !arrayFieldIncludes(document.frontmatter, 'tags', filters.tag)) return false
  if (filters.keyword && !arrayFieldIncludes(document.frontmatter, 'keywords', filters.keyword)) return false
  return true
}

const hasFilters = (filters: XDocsMetaFilters): boolean =>
  Boolean(filters.owner || filters.tag || filters.keyword)

const createMetaFilters = (options: XDocsMetaScanOptions): XDocsMetaFilters => ({
  ...(options.owner ? { owner: options.owner } : {}),
  ...(options.tag ? { tag: options.tag } : {}),
  ...(options.keyword ? { keyword: options.keyword } : {}),
})

const arrayFieldIncludes = (frontmatter: XDocsFrontmatter | null, key: string, expected: string): boolean => {
  const value = frontmatter?.[key]
  if (!Array.isArray(value)) return false
  return value.some((entry) => typeof entry === 'string' && sameValue(entry, expected))
}

const sameValue = (left: string | null | undefined, right: string): boolean =>
  typeof left === 'string' && left.toLowerCase() === right.toLowerCase()

const stringField = (frontmatter: XDocsFrontmatter | null, key: string): string | null => {
  const value = frontmatter?.[key]
  return typeof value === 'string' && value.length > 0 ? value : null
}

const walkMetadataDirectory = async (
  dir: string,
  config: XDocsConfig,
  onFile: (filePath: string) => Promise<void>,
): Promise<void> => {
  let entries

  try {
    entries = await readdir(dir, { withFileTypes: true })
  } catch {
    return
  }

  for (const entry of entries) {
    const fullPath = join(dir, entry.name)

    if (entry.isDirectory()) {
      if (isExcluded(entry.name, config.scan.exclude)) continue
      await walkMetadataDirectory(fullPath, config, onFile)
      continue
    }

    if (entry.isFile()) await onFile(fullPath)
  }
}

const isExcluded = (name: string, exclude: string[]): boolean =>
  exclude.includes(name) || name.startsWith('.')

const addDocumentReference = (
  documentsByDirectory: Map<string, MarkdownDocumentReference[]>,
  document: MarkdownDocumentReference,
): void => {
  const documents = documentsByDirectory.get(document.directory) ?? []
  documents.push(document)
  documentsByDirectory.set(document.directory, documents)
}

const isPlainMarkdownDocumentName = (name: string): boolean => {
  const lower = name.toLowerCase()
  return !name.includes('/') && !name.includes('\\') && lower.endsWith('.md') && lower !== 'xdocs.md' && !lower.endsWith('.xdocs.md')
}

File: internal/xdocs/meta.go
package xdocs

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/CGuiho/xdocs/internal/apperror"
	"github.com/CGuiho/xdocs/internal/config"
)

func ScanMetadata(cfg config.Config, options MetaOptions) (MetaResult, error) {
	target := cfg.CWD
	if options.TargetPath != "" {
		target = options.TargetPath
		if !filepath.IsAbs(target) {
			target = filepath.Join(cfg.CWD, target)
		}
	}
	target, _ = filepath.Abs(target)
	info, err := os.Stat(target)
	if err != nil {
		return MetaResult{}, apperror.New(apperror.Usage, fmt.Sprintf("metadata target does not exist: %s", slashRelative(cfg.CWD, target)))
	}
	if !info.IsDir() {
		return MetaResult{}, apperror.New(apperror.Usage, fmt.Sprintf("metadata target must be a directory: %s", slashRelative(cfg.CWD, target)))
	}

	result := MetaResult{
		Root:             cfg.CWD,
		TargetPath:       displayTarget(cfg.CWD, target),
		IncludeDocuments: options.IncludeDocuments,
		Strict:           options.Strict,
		Filters:          options.Filters,
		Descriptors:      []MetaDescriptor{},
		Errors:           []string{},
	}
	documents := map[string][]MarkdownDocument{}
	err = filepath.WalkDir(target, func(path string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			if entry != nil && entry.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if entry.IsDir() {
			if path != target && excluded(entry.Name(), cfg.Exclude) {
				return filepath.SkipDir
			}
			return nil
		}
		if IsDescriptor(path) {
			result.Descriptors = append(result.Descriptors, parseMetaDescriptor(path, cfg.CWD))
		} else if IsPlainMarkdown(path) {
			document := MarkdownDocument{Path: path, RelativePath: slashRelative(cfg.CWD, path), Directory: filepath.Dir(path), Name: filepath.Base(path)}
			documents[document.Directory] = append(documents[document.Directory], document)
		}
		return nil
	})
	if err != nil {
		return MetaResult{}, err
	}
	sort.Slice(result.Descriptors, func(i, j int) bool { return result.Descriptors[i].RelativePath < result.Descriptors[j].RelativePath })
	enrichMeta(result.Descriptors, documents, options.IncludeDocuments, cfg.CWD)
	result.Descriptors = filterDescriptors(result.Descriptors, options.Filters, options.IncludeDocuments)
	for _, descriptor := range result.Descriptors {
		for _, message := range descriptor.Errors {
			result.Errors = append(result.Errors, descriptor.RelativePath+": "+message)
		}
		for _, document := range descriptor.Documents {
			for _, message := range document.Errors {
				result.Errors = append(result.Errors, document.RelativePath+": "+message)
			}
		}
	}
	return result, nil
}

func parseMetaDescriptor(path, root string) MetaDescriptor {
	result := MetaDescriptor{
		Path:         path,
		RelativePath: slashRelative(root, path),
		Directory:    filepath.Dir(path),
		Documents:    []MetaDocument{},
		Errors:       []string{},
	}
	raw, ok, err := ReadFrontmatter(path)
	if err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("Read frontmatter: %v", err))
		return result
	}
	if !ok {
		result.Errors = append(result.Errors, "Missing YAML frontmatter.")
		return result
	}
	metadata, frontmatter, errors := parseMetadata(raw)
	result.Metadata = metadata
	result.Frontmatter = frontmatter
	result.Errors = append(result.Errors, errors...)
	if metadata != nil {
		result.Subject = metadata.Subject
	} else {
		result.Subject = stringValue(frontmatter, "subject")
	}
	result.Valid = metadata != nil && len(result.Errors) == 0
	return result
}

func enrichMeta(descriptors []MetaDescriptor, documents map[string][]MarkdownDocument, includeDocuments bool, root string) {
	byDirectory := map[string][]int{}
	for index := range descriptors {
		byDirectory[descriptors[index].Directory] = append(byDirectory[descriptors[index].Directory], index)
	}
	for _, indexes := range byDirectory {
		if len(indexes) <= 1 {
			continue
		}
		for _, index := range indexes {
			descriptors[index].Errors = append(descriptors[index].Errors, `Multiple xdocs descriptors found in this directory. Keep exactly one named "*.xdocs.md" file per directory.`)
		}
	}
	for index := range descriptors {
		descriptor := &descriptors[index]
		if descriptor.Metadata == nil {
			continue
		}
		actual := map[string]MarkdownDocument{}
		for _, document := range documents[descriptor.Directory] {
			actual[document.Name] = document
			if _, ok := descriptor.Metadata.Documents[document.Name]; !ok {
				descriptor.Errors = append(descriptor.Errors, fmt.Sprintf(`Undocumented Markdown document: "%s" must be listed in the "documents" metadata map.`, document.Name))
			}
		}
		for name := range descriptor.Metadata.Documents {
			if !validPlainMarkdownName(name) {
				descriptor.Errors = append(descriptor.Errors, fmt.Sprintf(`Invalid document entry: "%s" must be a sibling plain "*.md" filename, not an xdocs descriptor or path.`, name))
				continue
			}
			document, ok := actual[name]
			if !ok {
				descriptor.Errors = append(descriptor.Errors, fmt.Sprintf(`Missing Markdown document: "%s" is listed in metadata but does not exist beside the descriptor.`, name))
				continue
			}
			if includeDocuments {
				descriptor.Documents = append(descriptor.Documents, parseDocument(document.Path, root, descriptor.Metadata.Subject))
			}
		}
		sort.Slice(descriptor.Documents, func(i, j int) bool {
			return descriptor.Documents[i].RelativePath < descriptor.Documents[j].RelativePath
		})
		descriptor.Valid = descriptor.Metadata != nil && len(descriptor.Errors) == 0
	}
}

func filterDescriptors(descriptors []MetaDescriptor, filters Filters, includeDocuments bool) []MetaDescriptor {
	if filters.Owner == "" && filters.Tag == "" && filters.Keyword == "" {
		return descriptors
	}
	var result []MetaDescriptor
	for _, descriptor := range descriptors {
		matchedDocuments := []MetaDocument{}
		if includeDocuments {
			for _, document := range descriptor.Documents {
				if documentMatches(document, filters) {
					matchedDocuments = append(matchedDocuments, document)
				}
			}
		}
		descriptor.Documents = matchedDocuments
		if descriptorMatches(descriptor, filters) || len(matchedDocuments) > 0 {
			result = append(result, descriptor)
		}
	}
	return result
}

func descriptorMatches(descriptor MetaDescriptor, filters Filters) bool {
	if filters.Owner != "" && !strings.EqualFold(descriptor.Subject, filters.Owner) {
		return false
	}
	if descriptor.Metadata == nil {
		return filters.Owner == "" && filters.Tag == "" && filters.Keyword == ""
	}
	if filters.Tag != "" && !containsFold(descriptor.Metadata.Tags, filters.Tag) {
		return false
	}
	if filters.Keyword != "" && !containsFold(descriptor.Metadata.Keywords, filters.Keyword) {
		return false
	}
	return true
}

func documentMatches(document MetaDocument, filters Filters) bool {
	if filters.Owner != "" && !strings.EqualFold(document.Owner, filters.Owner) {
		return false
	}
	tags, _ := stringSlice(document.Frontmatter["tags"])
	keywords, _ := stringSlice(document.Frontmatter["keywords"])
	return (filters.Tag == "" || containsFold(tags, filters.Tag)) &&
		(filters.Keyword == "" || containsFold(keywords, filters.Keyword))
}

func containsFold(values []string, expected string) bool {
	for _, value := range values {
		if strings.EqualFold(value, expected) {
			return true
		}
	}
	return false
}

func displayTarget(root, path string) string {
	relative := slashRelative(root, path)
	if relative == "." || relative == "" {
		return "."
	}
	return relative
}
